nlp/gender_debiasing_cda/gender_debiasing_cda.ipynb

```diff
@@ -0,0 +1,1032 @@
+        "\n",
+        "We pre-downloaded the dataset from the Github page. If you are unable to do so, feel free to just load in our preprocessed dataset in the section **Load in the data**."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "CdIlM9fPT4SS"
+      },
+      "outputs": [],
+      "source": [
+        "wiki_bio_sent = []\n",
+        "wiki_bio_nb = []\n",
+        "wiki_bio_title = []\n",
+        "wiki_bio_id = []\n",
+        "\n",
+        "with open('wikipedia-biography-dataset/test/test.sent', 'r') as f:\n",
+        "    wiki_bio_sent = f.read().splitlines() \n",
+        "\n",
+        "with open('wikipedia-biography-dataset/test/test.nb', 'r') as f:\n",
+        "    wiki_bio_nb = f.read().splitlines()\n",
+        "\n",
+        "with open('wikipedia-biography-dataset/test/test.title', 'r') as f:\n",
+        "    wiki_bio_title = f.read().splitlines() \n",
+        "\n",
+        "with open('wikipedia-biography-dataset/test/test.id', 'r') as f:\n",
+        "    wiki_bio_id = f.read().splitlines() "
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "zvu9vWdiT4SZ"
+      },
+      "source": [
+        "It's in a bit of a weird format, so we'll have to merge it together"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "KldSaS8bT4Sc"
+      },
+      "outputs": [],
+      "source": [
+        "i = 0\n",
+        "wiki_bio_sent_grouped = []\n",
+        "for nr in wiki_bio_nb:\n",
+        "    nr_int = int(nr)\n",
+        "    wiki_sent_slice = wiki_bio_sent[i:i+nr_int]\n",
+        "    wiki_bio_sent_grouped.append(' '.join(wiki_sent_slice))\n",
+        "    i+=nr_int"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "OZC3N9cUT4Sf"
+      },
+      "outputs": [],
+      "source": [
+        "df_data = pd.DataFrame({\n",
+        "    'text': wiki_bio_sent_grouped,\n",
+        "})"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "HC1Qn5AUT4Si"
+      },
+      "outputs": [],
+      "source": [
+        "# Create a HF dataset object\n",
+        "hf_data = Dataset.from_pandas(df_data)\n",
+        "hf_data.save_to_disk('./data/hf_data')"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "RpigL1sxuelX"
+      },
+      "source": [
+        "## Counterfactual Data Augmentation (CDA) 🧑 👩\n",
+        "\n",
+        "To reduce gender bias in our dataset, we can use CDA, a technique that replaces every occurrence of a gendered word in the original corpus with its dual. For example:\n",
+        "- 'he' is replaced by 'she'\n",
```

Comment:
Doesn't this introduce other problems by not having all permutations? When every instance of "She goes out with her sisters." only turns into "He goes out with his brothers." but not into "She goes out with her brothers / he goes out with her sisters", we are discarding some easy possibilities to get rid of more bias. Or do I misunderstand line 183?

Reply:
That would indeed add an extra level of granularity. The problem here is that you would need to do coreference resolution first to not have inconsistencies in some permutations.

`She goes out with her sisters` cannot be augmented to `She goes out with his brothers`

A full swap is a bit crude indeed, but already seems to introduce a working hard counterpart of the initial dataset in an easy way

Reply:
But added an extra line on this